[WIP] Static analysis with PHPStan

ondrejmirtes opened the pull request:
Continuing the work from #9898. This is PR to integrate [PHPStan](https://github.com/phpstan/phpstan) into the CI build of CakePHP.

It's a work in progress since there are quite a lot of errors in the build log. `@property` annotations are already supported by PHPStan, but it still found some undefined properties.

Even after merging this PR, I expect to continue to work with you on increasing the level of strictness of the analysis and also on updating to newer versions of PHPStan :)

I'd like one of the maintainers to go through the error log and fix obvious errors in the code. We can then work together on the other, more questionable, errors.

ondrejmirtes committed:
Static analysis with PHPStan

ondrejmirtes commented:
Also, the mergable version of this PR will use a stable version of PHPStan. I'm waiting with the release in case we find something to fix in it.

ADmad commented on .travis.yml:
```diff
@@ -35,7 +35,10 @@ matrix:
   include:
     - php: 7.0
-      env: PHPCS=1 DEFAULT=0
+      env: PHPCS=1 DEFAULT=0 PHPSTAN=1
```
This will make phpstan check run along with the phpcs check. It would be better to have the phpstan check in a job of it's own so that it can be added to "allow_failures" for now. It will be a while before we can make all required changes to code to make it pass.

ondrejmirtes replied:
Should be better now.

dereuromark commented on .travis.yml:
```diff
@@ -37,6 +37,12 @@ matrix:
     - php: 7.0
       env: PHPCS=1 DEFAULT=0
 
+    - php: 7.0
```
Do we need both, or should we just have 7.1 here (and also in allowed failure list then). The result should be the same pretty much either way, right?

ondrejmirtes replied:
It will differ in case that something in PHP changed because I use reflection of the current version.

ondrejmirtes replied:
I think it's a good idea to run both.

ADmad replied:
Yes testing against PHP 7 and 7.1 doesn't seem necessary especially when current CakePHP version support PHP 5.5+ and we don't use any PHP 7 features.

dereuromark replied:
I am just a bit worried about the average build time then as it will further increase with each specific run. But I guess for the time being it doesnt hurt too much and those phpstan builds are probably rather fast.

dereuromark replied:
Let's get one (7.0 then maybe) to work first, then lets think about adding more.

ADmad commented:
> Also, the mergable version of this PR will use a stable version of PHPStan. I'm waiting with the release in case we find something to fix in it.

Using the master branch for a while is not a problem, at least until the phpstan job is under "allow_failures". That way you too have a sizeable codebase to test against when making updates to phpstan :smile:

dereuromark commented on .travis.yml:
```diff
@@ -45,6 +51,7 @@ matrix:
 
   allow_failures:
     - php: hhvm
+    - env: PHPSTAN=1
```
For a failure list to work you also need to include the php version each time afaik.

ondrejmirtes replied:
I'm not sure, we'll see :)

ADmad replied:
@ondrejmirtes https://travis-ci.org/cakephp/cakephp/builds/187472461  so @dereuromark is right.

ADmad commented:
@ondrejmirtes Houston we have a problem https://travis-ci.org/cakephp/cakephp/jobs/187472483#L628

ondrejmirtes commented:
@ADmad That's actually a problem with the code of CakePHP :) You cannot use `Exception` because `Cake\Database\Exception` already exists. So when you reference an `Exception` in the Connection class, PHP does not know which exception you mean. I will add a commit to this PR fixing that.

ondrejmirtes committed:
Fixed ambiguous exception in Cake\Database\Connection

ADmad commented:
> That's actually a problem with the code of CakePHP :)

Well the code works perfect fine in it's current state :)

ondrejmirtes commented:
@ADmad Only thanks to a coincidence. If you `require` the Exception first and the Connection second, you will end up with this error.

ondrejmirtes committed:
Fixed calling parent constructors

Some constructors could be even deleted along with duplicated protected properties.

ondrejmirtes committed:
Fixed invalid typehints

ondrejmirtes committed:
Removed unused constructor parameters

dereuromark commented:
Looking pretty good already

A few simple ones like

    Call to method Cake\ORM\Association\BelongsToMany::_selectableoptions()                                    
         with incorrect case: _selectableOptions      

are relatively easy to fix, I can take care of those after merge, as well.

dereuromark commented on src/Database/Connection.php:
```diff
@@ -185,7 +184,7 @@ public function connect()
 
             return true;
-        } catch (Exception $e) {
+        } catch (\Exception $e) {
```
Why is this necessary again? It should be OK to not use FQCN inside code and always include them via use statements.

ondrejmirtes replied:
It's basically this problem:

https://gist.github.com/cvuorinen/e7a19101d40c9fa0a959
https://3v4l.org/cphKu

dereuromark commented on src/Http/Client/Auth/Digest.php:
```diff
@@ -36,9 +36,8 @@ class Digest
-     * @param array|null $options Options list.
      */
-    public function __construct(Client $client, $options = null)
+    public function __construct(Client $client)
```
Not sure we can change the signature in a bugfix release on master, for reasons of BC.

ondrejmirtes replied:
I think this change is fine. The second parameter was optional and not used anyway, so it cannot break any code.

ondrejmirtes replied:
I originally removed the Response parameter from SessionStorage but returned it later because that change could actually break something.

dereuromark commented on src/View/Form/NullContext.php:
```diff
@@ -36,9 +36,8 @@ class NullContext implements ContextInterface
-     * @param array $context Context info.
      */
-    public function __construct(Request $request, array $context)
+    public function __construct(Request $request)
```
Same here..

ondrejmirtes replied:
The problem is that the parameter may be unused because of two reasons:

* It's really not supposed to be used and should be removed.
* It's forgotten to be used and the code should be fixed. Like in this code sample:

```
private $foo;

private $bar;

public function __construct($foo, $bar)
{
    $this->foo = $foo;
}
```

This check actually found a few of these situations in our codebase so I'm convinced it should be part of the default level 0.

dereuromark commented:
As for the Type changes, I am 1000% with you on those, but that was a constant pull and push here regarding some of the other core devs wanting to have those bad code blocks in there for theoretical future refactoring - see https://github.com/cakephp/cakephp/pull/9907
So it might be best for you not to open up that can of worms again ;)

ADmad commented on src/Auth/DigestAuthenticate.php:
```diff
@@ -81,16 +81,14 @@ class DigestAuthenticate extends BasicAuthenticate
         ]);
 
-        $this->config($config);
+        parent::__construct($registry, $config);
```
Seeing this change I remember I have already fixed this and few other things on the `3.next` branch. It would be best if this PR was targeted to `3.next` instead of `master`.

ADmad commented:
@ondrejmirtes Would you mind opening a new pull request against the `3.next` with just modifications to `.travis.yml` and the `phpstan.neon` file?  We can then review the report and make code changes or add ignores as required.

dereuromark commented:
@ADmad It does contain some fixes that could be valuable for the current master. What do you think, can be still merge those?

ADmad commented:
Given that 3.4 is round the corner I don't think it's worth the effort.

ondrejmirtes commented:
OK, I don't know your Git workflow so I assumed I should open the PR on master. I will make a new one and we'll see how it builds :)

ADmad commented:
> OK, I don't know your Git workflow so I assumed I should open the PR on master.

No problem, my fault for noticing the target branch sooner.

ondrejmirtes commented:
Continuing in #9943 :)
